miniml/compiler/test/patterns.ml

```diff
@@ -0,0 +1,28 @@
+open Lib
+
+let () = heading "Pattern-matching:"
+
+let () =
+  print_int (match [] with [] -> 2 | x :: l -> 3)
+
+let () =
+  print_int (match 1 :: [] with
+    | [] -> 2 (* note: leading bar *)
+    | _ :: _ -> 3)
+
+let test_function = function
+  | [] -> 2
+  | x :: _ -> x + 1 (* note: one of the pattern arguments is a wildcard *)
+
+let () =
+  print_int (test_function (3 :: []))
+
+type 'a tree =
+| Leaf of 'a
+| Node of 'a tree * 'a tree
+
+let () = print_int (match Node (Leaf 1, Leaf 2) with
+  | Leaf _ -> 4
+  | Node _ -> 5 (* note: a single wildcard for several arguments *)
+
+let () = finish ()
```

Comment:
The reference file for this file is empty.

Reply:
The reason why this reference file is empty is that running the corresponding test silently produces an empty output. When we look at the logs in more details (but the logs are a bit noisy due to Makefile noise and the compile.scm warning about `map`), we see the following:

```
Syntax error: unexpected token : LET #f (at line 27, column 1)
```

There is a syntax error in the test file caused by a missing parenthesis.

The correct thing to do would be to fail the test noisily here. By default make will complain if one of the commands returns an error code, so we should try to error harder on parse errors.

Reply:
(I have not fixed this one yet.)

Reply:
Done: now `errorp` will exit with a non-zero error code. This may change the behavior of the compiler in some cases (in multifile scenarios it might have returned `#nil` and continued to compile the other files), but I think that stopping with a hard error is a nicer behavior anyway.